p2p/chain.py

```diff
@@ -741,6 +633,144 @@ def _is_receipts_empty(header: BlockHeader) -> bool:
+        skip: int, reverse: bool, logger: logging.Logger, token: CancelToken) -> List[BlockHeader]:
+    """
+    Lookup :max_headers: headers starting at :block_number_or_hash:, skipping :skip: items between
+    each, in reverse order if :reverse: is True.
+    """
+    try:
+        block_numbers = await _get_block_numbers_for_request(
+            headerdb, block_number_or_hash, max_headers, skip, reverse, token)
+    except HeaderNotFound:
+        logger.debug(
+            "Peer requested starting header %r that is unavailable, returning nothing",
+            block_number_or_hash)
+        block_numbers = []
+
+    headers = [header async for header in _generate_available_headers(
+        headerdb, block_numbers, logger, token)]
+    return headers
+
+
+async def handle_get_block_bodies(
+        chaindb: 'AsyncChainDB', peer: ETHPeer, block_hashes: List[Hash32],
+        logger: logging.Logger, token: CancelToken) -> None:
+    bodies = []
+    # Only serve up to eth.MAX_BODIES_FETCH items in every request.
+    for block_hash in block_hashes[:eth.MAX_BODIES_FETCH]:
+        try:
+            header = await wait_with_token(
+                chaindb.coro_get_block_header_by_hash(block_hash),
+                token=token)
+        except HeaderNotFound:
+            logger.debug("%s asked for block we don't have: %s", peer, block_hash)
+            continue
+        transactions = await wait_with_token(
+            chaindb.coro_get_block_transactions(header, BaseTransactionFields),
+            token=token)
+        uncles = await wait_with_token(
+            chaindb.coro_get_block_uncles(header.uncles_hash),
+            token=token)
+        bodies.append(BlockBody(transactions, uncles))
+    peer.sub_proto.send_block_bodies(bodies)
+
+
+async def handle_get_receipts(
+        chaindb: 'AsyncChainDB', peer: ETHPeer, block_hashes: List[Hash32],
+        logger: logging.Logger, token: CancelToken) -> None:
+    receipts = []
+    # Only serve up to eth.MAX_RECEIPTS_FETCH items in every request.
+    for block_hash in block_hashes[:eth.MAX_RECEIPTS_FETCH]:
+        try:
+            header = await wait_with_token(
+                chaindb.coro_get_block_header_by_hash(block_hash),
+                token=token)
+        except HeaderNotFound:
+            logger.debug(
+                "%s asked receipts for block we don't have: %s", peer, block_hash)
+            continue
+        block_receipts = await wait_with_token(
+            chaindb.coro_get_receipts(header, Receipt),
+            token=token)
+        receipts.append(block_receipts)
+    peer.sub_proto.send_receipts(receipts)
+
+
+async def handle_get_node_data(
+        chaindb: 'AsyncChainDB', peer: ETHPeer, node_hashes: List[Hash32],
+        logger: logging.Logger, token: CancelToken) -> None:
+    nodes = []
+    # Only serve up to eth.MAX_STATE_FETCH items in every request.
+    for node_hash in node_hashes[:eth.MAX_STATE_FETCH]:
```

Comment:
My gut says that the truncation of which hashes we retrieve should happen at a higher level where this function is actually called, and that this function should blindly return all of the data that was requested.

I think this is a pretty minor thing, but architecturally, I think it is more correct.

- simplifies this function to be less complex.
- separates different classes of business logic.
- allows this function to be used for larger retrieval sizes (if that ever becomes a requirement)

Reply:
That makes sense, but this is where we call the sub-proto method that actually sends the data, so would need to extract this part into a separate unit that just returns the trie nodes. It'd also have to become a generator otherwise malicious peers could easily DoS us as we'd be always retrieving all the data from the DB, regardless of the limit.

How would you feel about just adding a new `limit` argument to the `handle_*()` methods for now, and having them default to the current values we use? That way we don't need to break these into even smaller methods (which is not necessary yet), but we keep them flexible and easy to refactor should it become necessary

Reply:
My intention was not to truncate the return value, but to truncate `node_hashes` before it was passed into this function.

```python
# at the call-site
trie_data = self._handler.handle_get_node_data(..., node_hashes=requested_hashes[:MAX_STATE_FETCH])
```

I think I prefer this over the `limit` argument, but both would be fine.

Reply:
Oh, I see! Went with your proposed solution